Client/Frontend/Share/ShareExtensionHelper.swift

```diff
@@ -54,27 +66,69 @@ class ShareExtensionHelper: NSObject {
-            if !completed {
-                completionHandler(completed, activityType)
-                return
-            }
-            // Bug 1392418 - When copying a url using the share extension there are 2 urls in the pasteboard.
-            // This is a iOS 11.0 bug. Fixed in 11.2
-            if UIPasteboard.general.hasURLs, let url = UIPasteboard.general.urls?.first {
-                UIPasteboard.general.urls = [url]
-            }
+        activityViewController.completionWithItemsHandler = { [weak self] activityType, completed, returnedItems, activityError in
+            guard let self = self else { return }
             
-            if self.isPasswordManagerActivityType(activityType.map { $0.rawValue }) {
-                if let logins = returnedItems {
-                    self.fillPasswords(logins as [AnyObject])
+            if self.shareActivityType(activityType.map { $0.rawValue }) != .iBooks { //!self.isOpenInIBooksActivityType(activityType.map { $0.rawValue }) {
+                if !completed {
+                    completionHandler(completed, activityType, nil)
+                    return
+                }
+                // Bug 1392418 - When copying a url using the share extension there are 2 urls in the pasteboard.
+                // This is a iOS 11.0 bug. Fixed in 11.2
+                if UIPasteboard.general.hasURLs, let url = UIPasteboard.general.urls?.first {
+                    UIPasteboard.general.urls = [url]
+                }
+                
+                if self.shareActivityType(activityType.map { $0.rawValue }) != .password { //self.isPasswordManagerActivityType(activityType.map { $0.rawValue }) {
+                    if let logins = returnedItems {
+                        self.fillPasswords(logins as [AnyObject])
+                    }
                 }
-            }
 
-            completionHandler(completed, activityType)
+                completionHandler(completed, activityType, nil)
+            } else {
+                self.writeWebPagePDFDataToURL { url, error in
+                    completionHandler(completed, activityType, url)
+                }
+            }
         }
+        
         return activityViewController
     }
+    
+    func writeWebPagePDFDataToURL(_ completion: @escaping (URL?, Error?) -> Void) {
+        #if compiler(>=5.3)
+        if #available(iOS 14.0, *), let webView = selectedTab?.webView, selectedTab?.temporaryDocument == nil {
+            
+            webView.createPDF { result in
+                dispatchPrecondition(condition: .onQueue(.main))
+                
+                switch result {
+                case .success(let data):
+                    let validFilenameSet = CharacterSet(charactersIn: ":/")
+                        .union(.newlines)
+                        .union(.controlCharacters)
+                        .union(.illegalCharacters)
```

Comment:
@soner-yuksel - do you have examples of filenames in which `controlCharacters` and `illegalCharacters` are part of the filename?

Reply:
This is for creating a file name for the pdf we want to save. So If it is extracting the string components without Illegal Unicode Characters (Cn) and Illegal (Cc, Cf) or New Line.

It is creating this name from the webview title so actually it can have anything there thats why.

I dont know have an example but I guess we can try to upload a pdf to a website with title including any character if it is necessary.

Reply:
As a defense in depth strategy, I think we can remove the `controlCharacters` and `illegalCharacters` from the `validFileNameSet` to avoid any path-traversal issues.